planning/planning_validator/autoware_planning_validator_collision_checker/src/collision_checker.cpp

```diff
@@ -0,0 +1,523 @@
+      params_.pointcloud.min_height,
+      context_->vehicle_info.vehicle_height_m + params_.pointcloud.height_buffer);
+    filter.filter(*filtered_pointcloud);
+  }
+
+  if (filtered_pointcloud->empty()) return;
+
+  {
+    geometry_msgs::msg::TransformStamped transform_stamped;
+    try {
+      transform_stamped = context_->tf_buffer.lookupTransform(
+        "map", input->header.frame_id, input->header.stamp, rclcpp::Duration::from_seconds(0.1));
+    } catch (tf2::TransformException & e) {
+      RCLCPP_WARN(logger_, "no transform found for pointcloud: %s", e.what());
+    }
+
+    Eigen::Affine3f isometry = tf2::transformToEigen(transform_stamped.transform).cast<float>();
+    autoware_utils::transform_pointcloud(*filtered_pointcloud, *filtered_pointcloud, isometry);
+  }
+
+  {
+    const auto & p = params_.pointcloud;
+    pcl::VoxelGrid<pcl::PointXYZ> filter;
+    filter.setInputCloud(filtered_pointcloud);
+    filter.setLeafSize(p.voxel_grid_filter.x, p.voxel_grid_filter.y, p.voxel_grid_filter.z);
+    filter.filter(*filtered_pointcloud);
+  }
+}
+
+bool CollisionChecker::check_collision(
+  const TargetLanelets & target_lanelets, const PointCloud::Ptr & filtered_point_cloud,
+  const rclcpp::Time & time_stamp)
+{
+  bool is_safe = true;
+
+  auto ego_object_overlap_time =
+    [](const double object_ttc, const std::pair<double, double> & ego_time) {
+      if (object_ttc > ego_time.second) return object_ttc - ego_time.second;
+      if (object_ttc < ego_time.first) return ego_time.first - object_ttc;
+      return 0.0;
+    };
+
+  for (const auto & target_lanelet : target_lanelets) {
+    if (target_lanelet.lanelets.empty()) continue;
+
+    const auto pcd_object = get_pcd_object(time_stamp, filtered_point_cloud, target_lanelet);
+    if (!pcd_object.has_value()) continue;
+
+    auto update_object = [&](PCDObject & object, const PCDObject & new_data) {
+      const auto dt = (new_data.last_update_time - object.last_update_time).seconds();
+      const auto dl = object.distance_to_overlap - new_data.distance_to_overlap;
+      if (dt < 1e-6) return;  // too small time difference, skip update
+
+      object.track_duration += dt;
+      const auto max_accel = 30.0;
+      const auto raw_velocity = dl / dt;
+      const bool is_reliable = object.track_duration > params_.pointcloud.observation_time;
+      if (is_reliable && std::abs(raw_velocity - object.velocity) / dt > max_accel) {
+        object.velocity = 0.0;        // too high acceleration, reset velocity
+        object.track_duration = 0.0;  // reset track duration
+      } else {
+        object.velocity = autoware::signal_processing::lowpassFilter(
+          raw_velocity, object.velocity, 0.5);  // apply low-pass filter to velocity
+      }
+      object.last_update_time = new_data.last_update_time;
+      object.pose = new_data.pose;
+      object.distance_to_overlap = new_data.distance_to_overlap;
+      object.ttc = object.distance_to_overlap / object.velocity;
```

Comment:
Given that point cloud processing can introduce a delay of several hundred milliseconds, it might be advisable to implement delay compensation. Without such compensation, the TTC could be overestimated. However, it's also acceptable to handle this in a separate PR after reviewing the results of the evaluation experiment.